### DISCUSSION TITLE
Integration points stresses and volumes

### DISCUSSION THREAD

**Discussion by FrancescoSausto** [pyansys/pymapdl-reader#152] [Q&A] [answered]
Dear all,
I'm using PyMAPDL Reader for integrating several tools that I developed with Ansys. First of all I want to congratulate with the authors for the great work. The tool works great and it is really fast in opening big results file.

I have a question, is there some possibility to read also the stresses at the integration points of the elements and their associated volumes? I checked the documentation and it seems to me that the only stress values that can be read are nodal and at the element's centroid... maybe I missed the IPs results function.

Thanks in advance for the helping!

Francesco

**Comment by pmaroneh** [comment, marked as answer]
Hi @FrancescoSausto , thank you for the nice feedback. 
Stress data is saved at the nodes and can be postprocessed as unaveraged values (one value per node per element - PLESOL method), averaged values (one value per node - PRNSOL method ), or elemental mean values (ETABLE method). 
To get the stress at integration points, you can use mapdl.eresx('no') option : during the solve, the derived results (stresses, strains, etc) that are initially calculated at integration points will be copied to the nodes instead of being extrapolated.
Element volume can be obtained in the results, usually through etab,volu,volu.

**Comment by FrancescoSausto** [comment]
Hi @pmaroneh, thank you for the detailed and quick answer to my question. 
As I can understand your suggestion requires to run MAPDL via the python tool PyMAPDL and then request to the solver the informations I need (i.e. stress and volumes of the integration points). Actually I was interested in only reading these results from the RST file; somethimes I only have this database without the analysis hence it would be diffucult to me to run MAPDL. I'm not an expert on the topic, but I'm quite sure that the integration point results are available in the RST file, since when I required them "manually" from the workbench mechanical the simulation is not run again. 

Is there a way to ask for IPs results directly with PyMAPDL Reader? Or I need to use directly PyMAPDL as you suggested? 

Thank you so much, I really appreciete your availability!

Francesco

**Reply by pmaroneh** [reply]
Hi @FrancescoSausto . My suggestion indeed involves re-running the model and to copy (instead of extrapolating) the derived results from the integration points to the nodes. The results are stored at the nodes, we don't store stress/strain at every integration point since we don't visualize the integration points and doing so would lead to very large result file sizes. I don't believe WB Mechanical offers the possibility to post-process at the integration points either; which method have you used there?

**Reply by FrancescoSausto** [reply]
Hi @pmaroneh, thank you for the suggestions. I exported the results manually from workbench mechanical; once I chose the results to visualize I set them as "unaveraged" in the "integration point results" box. I think that I will go with simply considering the results at the elements' centroid with their associated volume. 

Thank you very much for helping! This discussion was very interesting!

Francesco

**Reply by pmaroneh** [reply]
Hi @FrancescoSausto I'm happy to help here. The unaveraged display option in Mechanical does not return the values at the integration point. It returns values at the node in unaveraged display (one value per node per element). 
Excuse my horrible drawings below but maybe they will help to fully understand. 
When solving for a static structural analysis we are resolving KU = F where U is the unknown displacement.
Once U has been found at the nodes we calculate derived values (stresses, strains) at the integration points. Then, as we don't display the integration points, the derived values are taken back to the nodes. The picture below illustrates what happens for one element : 
![image](https://user-images.githubusercontent.com/76440649/191913795-bd98ed9f-cfa3-4585-a7e0-ee3f165fff9f.png)
Depending on the situation the values are either interpolated (when at the element level the material is linear or when ERESX,YES is used) or copied (when at the element level nonlinearities are present or when ERESX,NO is used).
In the end, each node receives several values, one for each element they are connected to . The below images illustrates the case of a node being part of four elements : 
![image](https://user-images.githubusercontent.com/76440649/191914262-9de9958e-93b8-4689-9268-b0649563b47c.png)
The unaveraged option says we keep the four values (if you export the values from Mechanical to an Excel file you will see the same node number appear several times), the averaged option says we average the values (only one value per node).
If your mesh is fine enough both of these results should be similar.
